packages/forms-web-app/src/dynamic-forms/validator/date-validator.js

```diff
@@ -0,0 +1,153 @@
+	/**
+	 * creates an instance of a DateValidator
+	 * @param {string} [inputLabel] - string representing the field as displayed on the UI as part of an error message
+	 * @param {Object} [errorMessage] - object containing custom error messages to show on validation failure
+	 */
+	constructor(inputLabel, errorMessages) {
+		super();
+
+		const defaultErrorMessages = this.#getDefaultErrorMessages(inputLabel);
+
+		this.emptyErrorMessage =
+			errorMessages?.emptyErrorMessage ?? defaultErrorMessages.emptyErrorMessage;
+		this.noDayErrorMessage =
+			errorMessages?.noDayErrorMessage ?? defaultErrorMessages.noDayErrorMessage;
+		this.noMonthErrorMessage =
+			errorMessages?.noMonthErrorMessage ?? defaultErrorMessages.noMonthErrorMessage;
+		this.noYearErrorMessage =
+			errorMessages?.noYearErrorMessage ?? defaultErrorMessages.noYearErrorMessage;
+		this.noDayMonthErrorMessage =
+			errorMessages?.noDayMonthErrorMessage ?? defaultErrorMessages.noDayMonthErrorMessage;
+		this.noDayYearErrorMessage =
+			errorMessages?.noDayYearErrorMessage ?? defaultErrorMessages.noDayYearErrorMessage;
+		this.noMonthYearErrorMessage =
+			errorMessages?.noMonthYearErrorMessage ?? defaultErrorMessages.noMonthYearErrorMessage;
+		this.invalidDateErrorMessage =
+			errorMessages?.invalidDateErrorMessage ?? defaultErrorMessages.invalidDateErrorMessage;
+		this.invalidYearErrorMessage =
+			errorMessages?.invalidYearErrorMessage ?? defaultErrorMessages.invalidYearErrorMessage;
+		this.futureDateErrorMessage =
+			errorMessages?.futureDateErrorMessage ?? defaultErrorMessages.futureDateErrorMessage;
+	}
+
+	/**
+	 * validates the response body, checking the values sent for the date are valid
+	 * @param {Question} questionObj
+	 */
+	validate(questionObj) {
+		const fieldName = questionObj.fieldName;
+		const dayInput = `${fieldName}_day`;
+		const monthInput = `${fieldName}_month`;
+		const yearInput = `${fieldName}_year`;
+
+		return [
+			// check all or some date inputs are not empty
+			body(dayInput)
+				.notEmpty()
+				.withMessage((_, { req }) => {
+					if (!req.body[monthInput] && !req.body[yearInput]) {
+						return this.emptyErrorMessage;
+					}
+
+					if (!req.body[monthInput] && req.body[yearInput]) {
+						return this.noDayMonthErrorMessage;
+					}
+
+					if (req.body[monthInput] && !req.body[yearInput]) {
+						return this.noDayYearErrorMessage;
+					}
+
+					return this.noDayErrorMessage;
+				}),
+			body(monthInput)
+				.notEmpty()
+				.withMessage((_, { req }) => {
+					if (req.body[dayInput] && !req.body[yearInput]) {
+						return this.noMonthYearErrorMessage;
+					}
+					if (req.body[dayInput]) {
+						return this.noMonthErrorMessage;
+					}
+				}),
+			body(yearInput)
+				.notEmpty()
+				.withMessage((_, { req }) => {
+					if (req.body[dayInput] && req.body[monthInput]) return this.noYearErrorMessage;
+				}),
+
+			// check date values entered are within valid ranges
+			body(dayInput)
+				.isInt({ min: 1, max: 31 })
```

Comment:
I'd consider a utility method like so
`var isValid = require('date-fns/isValid')
isValidWrapper(year = 2000, month = 1, day = 1)
{
isValid(new Date(year, month, 1)) : true ? throw InvalidMonthError("Month is invalid month " + month);
}
// Then we can do, which will work on any day, we can do the same to check the month and year on their own above
if (isValidWrapper(day = value) {
						const yearNum = parseInt(req.body[yearInput], 10);
						return yearNum && isLeapYear(new Date(yearNum, 1, 1)) && value === 29;
					}

`